src/connector/src/sink/kafka.rs

```diff
@@ -0,0 +1,557 @@
+                    }
+                }
+            }
+        }
+        Err(err_placeholder)
+    }
+
+    fn gen_message_key(&self) -> String {
+        format!(
+            "{}-{}",
+            self.config.identifier,
+            self.in_transaction_epoch.unwrap()
+        )
+    }
+
+    async fn append_only(&self, chunk: StreamChunk, schema: &Schema) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            if op == Op::Insert {
+                let record = Value::Object(record_to_json(row, schema.fields.clone())?).to_string();
+                self.send(
+                    BaseRecord::to(self.config.topic.as_str())
+                        .key(self.gen_message_key().as_bytes())
+                        .payload(record.as_bytes()),
+                )
+                .await?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl Sink for KafkaSink {
+    async fn write_batch(&mut self, chunk: StreamChunk, schema: &Schema) -> Result<()> {
+        // when sinking the snapshot, it is required to begin epoch 0 for transaction
+        if self.in_transaction_epoch.unwrap() <= self.latest_success_epoch {
+            return Ok(());
+        }
+        if self.config.sink_type.as_str() == "append_only" {
+            self.append_only(chunk, schema).await
+        } else if self.config.sink_type.as_str() == "debezium" {
+            todo!()
+        } else {
+            unreachable!()
+        }
+    }
+
+    async fn begin_epoch(&mut self, epoch: u64) -> Result<()> {
+        self.in_transaction_epoch = Some(epoch);
+        if self.latest_success_epoch == 0 {
+            self.do_with_retry(|conductor| conductor.init_transaction())
+                .await
+                .map_err(SinkError::Kafka)?;
+        }
+
+        self.do_with_retry(|conductor| conductor.start_transaction())
+            .await
+            .map_err(SinkError::Kafka)?;
+        tracing::debug!("begin epoch {:?}", epoch);
+        Ok(())
+    }
+
+    async fn commit(&mut self) -> Result<()> {
+        self.do_with_retry(|conductor| conductor.flush()) // flush before commit
+            .await
+            .map_err(SinkError::Kafka)?;
+
+        self.do_with_retry(|conductor| conductor.commit_transaction())
+            .await
+            .map_err(SinkError::Kafka)?;
+        if let Some(epoch) = self.in_transaction_epoch.take() {
+            self.latest_success_epoch = epoch;
+        } else {
+            tracing::error!(
+                "commit without begin_epoch, last success epoch {:?}",
+                self.latest_success_epoch
+            );
+            return Err(SinkError::Kafka(KafkaError::Canceled));
+        }
+        Ok(())
+    }
+
+    async fn abort(&mut self) -> Result<()> {
+        self.do_with_retry(|conductor| conductor.abort_transaction())
+            .await
+            .map_err(SinkError::Kafka)
+    }
+
+    async fn take_snapshot(&self) -> Result<SinkState> {
```

Comment:
What's the difference between take_snapshot and commit?

Reply:
remove take_snapshot func, resolve now